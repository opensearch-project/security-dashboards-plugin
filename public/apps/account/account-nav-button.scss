/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

// stylelint-disable-next-line @osd/stylelint/no_modifying_global_selectors
.euiButtonEmpty.accountNavButton {
  border: 0;
}
