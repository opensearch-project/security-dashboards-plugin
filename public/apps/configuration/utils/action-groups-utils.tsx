/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { map } from 'lodash';
import { API_ENDPOINT_ACTIONGROUPS, CLUSTER_PERMISSIONS, INDEX_PERMISSIONS } from '../constants';
import { DataObject, ActionGroupItem, ActionGroupUpdate, ObjectsMessage } from '../types';
import { httpDelete, httpGet, httpPost } from './request-utils';
import { getResourceUrl } from './resource-utils';

export interface PermissionListingItem {
  name: string;
  type: 'Action group' | 'Single permission';
  reserved: boolean;
  allowedActions: string[];
  hasClusterPermission: boolean;
  hasIndexPermission: boolean;
}

export async function fetchActionGroups(http: HttpStart): Promise<DataObject<ActionGroupItem>> {
  const actiongroups = await httpGet<ObjectsMessage<ActionGroupItem>>({
    http,
    url: API_ENDPOINT_ACTIONGROUPS,
  });
  return actiongroups.data;
}

export function transformActionGroupsToListingFormat(
  rawData: DataObject<ActionGroupItem>
): PermissionListingItem[] {
  return map(rawData, (value: ActionGroupItem, key?: string) => ({
    name: key || '',
    type: 'Action group',
    reserved: value.reserved,
    allowedActions: value.allowed_actions,
    hasClusterPermission: value.type === 'cluster' || value.type === 'all',
    hasIndexPermission: value.type === 'index' || value.type === 'all',
  }));
}

export async function fetchActionGroupListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(await fetchActionGroups(http));
}

function getClusterSinglePermissions(): PermissionListingItem[] {
  return CLUSTER_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: true,
    hasIndexPermission: false,
  }));
}

function getIndexSinglePermissions(): PermissionListingItem[] {
  return INDEX_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: false,
    hasIndexPermission: true,
  }));
}

export async function getAllPermissionsListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return mergeAllPermissions(await fetchActionGroups(http));
}

export async function mergeAllPermissions(
  actionGroups: DataObject<ActionGroupItem>
): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(actionGroups)
    .concat(getClusterSinglePermissions())
    .concat(getIndexSinglePermissions());
}

export async function updateActionGroup(
  http: HttpStart,
  groupName: string,
  updateObject: ActionGroupUpdate
): Promise<ActionGroupUpdate> {
  return await httpPost({
    http,
    url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, groupName),
    body: updateObject,
  });
}

export async function requestDeleteActionGroups(http: HttpStart, groups: string[]) {
  for (const group of groups) {
    await httpDelete({ http, url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, group) });
  }
}
