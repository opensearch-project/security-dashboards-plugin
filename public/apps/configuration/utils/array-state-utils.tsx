/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { get, set, curry, PropertyName } from 'lodash';
import { Dispatch, SetStateAction } from 'react';

function resolveValue<T>(value: T | (() => T)) {
  return typeof value === 'function' ? (value as Function)() : value;
}

/**
 * Update an element in an sub array
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern"
 * @param newValue value to be updated
 *
 * e.g.
 *  currentState = [{a: 1}, {a: 2}]
 *  path = [0, 'a']
 *  newValue = 10
 * The newState will be [{a: 10}, {a: 2}]
 */
export function updateElementInArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  setStateCallback((prevState) => {
    const newState = [...prevState];
    set(newState, path, resolveValue(newValue));
    return newState;
  });
}

export const updateElementInArrayHandler = curry(updateElementInArray);

/**
 * Append an element to an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param newValue value to be updated
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  newValue = 3
 * The newState will be [1, 2, 3]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  newValue = 5
 * The newState will be [{a: [1, 2, 5]}, {a: [3, 4]}]
 */
export function appendElementToArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  const resolvedNewValue = resolveValue(newValue);
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      return [...prevState, resolvedNewValue];
    } else {
      const newArray = [...(get(prevState, path) as T[]), resolvedNewValue];
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}

/**
 * Remove an element from an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param index index of element to be removed
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  index = 0
 * The newState will be [2]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  index = 1
 * The newState will be [{a: [1]}, {a: [3, 4]}]
 */
export function removeElementFromArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  index: number
) {
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      const newState = [...prevState];
      newState.splice(index, 1);
      return newState;
    } else {
      const newArray = [...(get(prevState, path) as T[])];
      newArray.splice(index, 1);
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}
