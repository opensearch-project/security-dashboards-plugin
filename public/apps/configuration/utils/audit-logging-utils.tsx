/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { AuditLoggingSettings } from '../panels/audit-logging/types';
import { API_ENDPOINT_AUDITLOGGING, API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import { httpGet, httpPost } from './request-utils';

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await httpPost({ http, url: API_ENDPOINT_AUDITLOGGING_UPDATE, body: updateObject });
}

export async function getAuditLogging(http: HttpStart): Promise<AuditLoggingSettings> {
  const rawConfiguration = await httpGet<any>({ http, url: API_ENDPOINT_AUDITLOGGING });
  return rawConfiguration?.config;
}
