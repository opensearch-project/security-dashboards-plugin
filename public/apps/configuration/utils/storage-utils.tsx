/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { Toast } from '@elastic/eui/src/components/toast/global_toast_list';

/*
 * This file leverage storage for frontend data. Currently it is for cross page toasts only,
 * so session storage is in use (no need to hold the toast message to until next session)
 */

export function getValue<T>(category: string, key: string): T | null {
  const item = sessionStorage.getItem(`${category}::${key}`);
  if (!item) {
    return null;
  }
  return JSON.parse(item) as T;
}

export function setValue(category: string, key: string, value: any) {
  sessionStorage.setItem(`${category}::${key}`, JSON.stringify(value));
}

const TOAST_CATEGORY = 'toast';

export function setCrossPageToast(targetUrl: string, toast: Toast) {
  setValue(TOAST_CATEGORY, targetUrl, toast);
}

export function getAndClearCrossPageToast(targetUrl: string): Toast | null {
  const result = getValue<Toast>(TOAST_CATEGORY, targetUrl);
  setValue(TOAST_CATEGORY, targetUrl, null);
  return result;
}
