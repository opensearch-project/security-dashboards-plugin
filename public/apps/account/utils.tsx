/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { API_AUTH_LOGOUT, OPENID_AUTH_LOGOUT, SAML_AUTH_LOGOUT } from '../../../common';
import { API_ENDPOINT_ACCOUNT_INFO } from './constants';
import { AccountInfo } from './types';
import { httpGet, httpGetWithIgnores, httpPost } from '../configuration/utils/request-utils';
import { setShouldShowTenantPopup } from '../../utils/storage-utils';

export function fetchAccountInfo(http: HttpStart): Promise<AccountInfo> {
  return httpGet(http, API_ENDPOINT_ACCOUNT_INFO);
}

export async function fetchAccountInfoSafe(http: HttpStart): Promise<AccountInfo | undefined> {
  return httpGetWithIgnores<AccountInfo>(http, API_ENDPOINT_ACCOUNT_INFO, [401]);
}

export async function logout(http: HttpStart, logoutUrl?: string): Promise<void> {
  await httpPost(http, API_AUTH_LOGOUT);
  setShouldShowTenantPopup(null);
  // Clear everything in the sessionStorage since they can contain sensitive information
  sessionStorage.clear();
  // When no basepath is set, we can take '/' as the basepath.
  const basePath = http.basePath.serverBasePath ? http.basePath.serverBasePath : '/';
  const nextUrl = encodeURIComponent(basePath);
  window.location.href =
    logoutUrl || `${http.basePath.serverBasePath}/app/login?nextUrl=${nextUrl}`;
}

export async function samlLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  window.location.href = `${http.basePath.serverBasePath}${SAML_AUTH_LOGOUT}`;
}

export async function openidLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${OPENID_AUTH_LOGOUT}`;
}

export async function externalLogout(http: HttpStart, logoutEndpoint: string): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${logoutEndpoint}`;
}

export async function updateNewPassword(
  http: HttpStart,
  newPassword: string,
  currentPassword: string
): Promise<void> {
  await httpPost(http, API_ENDPOINT_ACCOUNT_INFO, {
    password: newPassword,
    current_password: currentPassword,
  });
}
