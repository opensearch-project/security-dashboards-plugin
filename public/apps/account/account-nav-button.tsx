/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiAvatar,
  EuiButtonEmpty,
  EuiContextMenuPanel,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHeaderSectionItemButton,
  EuiHorizontalRule,
  EuiListGroup,
  EuiListGroupItem,
  EuiPopover,
  EuiText,
} from '@elastic/eui';
import { CoreStart } from 'opensearch-dashboards/public';
import React, { useCallback } from 'react';
import { RoleInfoPanel } from './role-info-panel';
import { PasswordResetPanel } from './password-reset-panel';
import { TenantSwitchPanel } from './tenant-switch-panel';
import { ClientConfigType } from '../../types';
import { LogoutButton } from './log-out-button';
import { resolveTenantName } from '../configuration/utils/tenant-utils';
import { getShouldShowTenantPopup, setShouldShowTenantPopup } from '../../utils/storage-utils';
import { getDashboardsInfo } from '../../utils/dashboards-info-utils';

export function AccountNavButton(props: {
  coreStart: CoreStart;
  isInternalUser: boolean;
  username: string;
  tenant?: string;
  config: ClientConfigType;
  currAuthType: string;
}) {
  const [isPopoverOpen, setPopoverOpen] = React.useState<boolean>(false);
  const [modal, setModal] = React.useState<React.ReactNode>(null);
  const horizontalRule = <EuiHorizontalRule margin="xs" />;
  const username = props.username;
  const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(true);

  const showTenantSwitchPanel = useCallback(
    () =>
      setModal(
        <TenantSwitchPanel
          coreStart={props.coreStart}
          config={props.config}
          handleClose={() => {
            setModal(null);
          }}
          handleSwitchAndClose={() => {
            setModal(null);
            reloadAfterTenantSwitch();
          }}
          tenant={props.tenant!}
        />
      ),
    [props.config, props.coreStart, props.tenant]
  );
  React.useEffect(() => {
    const fetchData = async () => {
      try {
        setIsMultiTenancyEnabled(
          (await getDashboardsInfo(props.coreStart.http)).multitenancy_enabled
        );
      } catch (e) {
        // TODO: switch to better error display.
        console.error(e);
      }
    };

    fetchData();
  }, [props.coreStart.http]);

  // Check if the tenant modal should be shown on load
  if (getShouldShowTenantPopup()) {
    setShouldShowTenantPopup(false);
    showTenantSwitchPanel();
  }

  const contextMenuPanel = (
    <div style={{ maxWidth: '256px' }}>
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={null}>
          <EuiAvatar name={username} />
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiListGroup gutterSize="none">
            <EuiListGroupItem
              key="username"
              wrapText
              label={
                <EuiText size="s">
                  <h5>{username}</h5>
                </EuiText>
              }
            />
          </EuiListGroup>
          {isMultiTenancyEnabled && props.config.multitenancy.enabled && (
            <EuiListGroupItem
              color="subdued"
              key="tenant"
              label={
                <EuiText size="xs" id="tenantName">
                  {resolveTenantName(props.tenant || '', username)}
                </EuiText>
              }
            />
          )}
        </EuiFlexItem>
      </EuiFlexGroup>

      {horizontalRule}
      <EuiButtonEmpty
        data-test-subj="view-roles-and-identities"
        size="xs"
        onClick={() => setModal(<RoleInfoPanel {...props} handleClose={() => setModal(null)} />)}
      >
        View roles and identities
      </EuiButtonEmpty>
      {isMultiTenancyEnabled && props.config.multitenancy.enabled && (
        <>
          {horizontalRule}
          <EuiButtonEmpty data-test-subj="switch-tenants" size="xs" onClick={showTenantSwitchPanel}>
            Switch tenants
          </EuiButtonEmpty>
        </>
      )}
      {props.isInternalUser && (
        <>
          {horizontalRule}
          <EuiButtonEmpty
            data-test-subj="reset-password"
            size="xs"
            onClick={() =>
              setModal(
                <PasswordResetPanel
                  coreStart={props.coreStart}
                  username={props.username}
                  handleClose={() => setModal(null)}
                  logoutUrl={props.config.auth.logout_url}
                />
              )
            }
          >
            Reset password
          </EuiButtonEmpty>
        </>
      )}
      <LogoutButton
        authType={props.currAuthType}
        http={props.coreStart.http}
        divider={horizontalRule}
        logoutUrl={props.config.auth.logout_url}
      />
    </div>
  );
  return (
    <EuiHeaderSectionItemButton id="user-icon-btn">
      <EuiPopover
        data-test-subj="account-popover"
        id="actionsMenu"
        button={<EuiAvatar name={username} />}
        isOpen={isPopoverOpen}
        closePopover={() => {
          setPopoverOpen(false);
        }}
        panelPaddingSize="s"
        onClick={() => {
          setPopoverOpen((prevState) => !prevState);
        }}
      >
        <EuiContextMenuPanel>{contextMenuPanel}</EuiContextMenuPanel>
      </EuiPopover>
      {modal}
    </EuiHeaderSectionItemButton>
  );
}

export function reloadAfterTenantSwitch(): void {
  // the below portion is to clear URLs starting with 'lastUrl'
  // when switching tenants, the last URLs will be from the old tenancy therefore we need to remove these from sessionStorage.
  const lastUrls = [];
  const { sessionStorage } = window;

  for (let i = 0; i < sessionStorage.length; i++) {
    const key = sessionStorage.key(i);
    if (key?.startsWith('lastUrl')) {
      lastUrls.push(key);
    }
  }
  for (let i = 0; i < lastUrls.length; i++) {
    sessionStorage.removeItem(lastUrls[i]);
  }

  // rather than just reload when we switch tenants, we set the URL to the pathname. i.e. the portion like: '/app/dashboards'
  // therefore, the copied URL will now allow tenancy changes.
  window.location.href = window.location.pathname;
}
