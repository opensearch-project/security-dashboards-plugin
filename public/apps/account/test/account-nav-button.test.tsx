/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import { AccountNavButton, reloadAfterTenantSwitch } from '../account-nav-button';
import { getShouldShowTenantPopup } from '../../../utils/storage-utils';
import { getDashboardsInfo } from '../../../utils/dashboards-info-utils';
import { render, fireEvent } from '@testing-library/react';

jest.mock('../../../utils/storage-utils', () => ({
  getShouldShowTenantPopup: jest.fn(),
  setShouldShowTenantPopup: jest.fn(),
}));

jest.mock('../../../utils/dashboards-info-utils', () => ({
  getDashboardsInfo: jest.fn().mockImplementation(() => {
    return mockDashboardsInfo;
  }),
}));

const mockDashboardsInfo = {
  multitenancy_enabled: true,
  private_tenant_enabled: true,
  default_tenant: '',
};

describe('Account navigation button', () => {
  const mockCoreStart = {
    http: 1,
  };

  const config = {
    multitenancy: {
      enabled: true,
      tenants: {
        enable_private: true,
        enable_global: true,
      },
    },
    auth: {
      type: 'dummy',
    },
  };

  const userName = 'user1';

  let component;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');

  beforeEach(() => {
    useStateSpy.mockImplementation((init) => [init, setState]);
    (getDashboardsInfo as jest.Mock).mockImplementation(() => {
      return mockDashboardsInfo;
    });
    component = shallow(
      <AccountNavButton
        coreStart={mockCoreStart}
        isInternalUser={true}
        username={userName}
        tenant="tenant1"
        config={config as any}
        currAuthType={'dummy'}
      />
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    useStateSpy.mockRestore();
  });

  it('renders', () => {
    (getDashboardsInfo as jest.Mock).mockImplementationOnce(() => {
      return mockDashboardsInfo;
    });
    expect(component).toMatchSnapshot();
  });

  it('should set modal when show popup is true', () => {
    (getDashboardsInfo as jest.Mock).mockImplementation(() => {
      return mockDashboardsInfo;
    });
    (getShouldShowTenantPopup as jest.Mock).mockReturnValueOnce(true);
    shallow(
      <AccountNavButton
        coreStart={mockCoreStart}
        isInternalUser={true}
        username={userName}
        tenant="tenant1"
        config={config as any}
        currAuthType={'dummy'}
      />
    );
    expect(setState).toBeCalledTimes(1);
  });

  it('should set modal when click on "View roles and identities" button', () => {
    component.find('[data-test-subj="view-roles-and-identities"]').simulate('click');
    expect(setState).toBeCalledTimes(1);
  });

  it('should set modal when click on "Switch tenants" button', () => {
    component.find('[data-test-subj="switch-tenants"]').simulate('click');
    expect(setState).toBeCalledTimes(1);
  });

  it('should set modal when click on "Reset password" button', () => {
    component.find('[data-test-subj="reset-password"]').simulate('click');
    expect(setState).toBeCalledTimes(1);
  });

  it('should set isPopoverOpen to true when click on Avatar in header section', () => {
    component.find('[data-test-subj="account-popover"]').simulate('click');
    expect(setState).toBeCalledTimes(1);
  });
});

describe('Account navigation button, multitenancy disabled', () => {
  const mockCoreStart = {
    http: 1,
  };

  const config = {
    multitenancy: {
      enabled: false,
    },
    auth: {
      type: 'dummy',
    },
  };

  const userName = 'user1';
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');

  beforeEach(() => {
    useStateSpy.mockImplementation((init) => [init, setState]);
  });

  afterAll(() => {
    useStateSpy.mockRestore();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should not set modal when getShouldShowTenantPopup is false', () => {
    (getShouldShowTenantPopup as jest.Mock).mockReturnValueOnce(false);
    shallow(
      <AccountNavButton
        coreStart={mockCoreStart}
        isInternalUser={true}
        username={userName}
        config={config as any}
        currAuthType={'dummy'}
      />
    );
    expect(setState).toBeCalledTimes(0);
  });
});

describe('Shows tenant info when multitenancy enabled, and hides it if disabled', () => {
  test('Renders "switch-tenants" and "tenant-name" when multi-tenancy is enabled', () => {
    const props = {
      coreStart: {},
      isInternalUser: true,
      username: 'example_user',
      tenant: 'example_tenant',
      config: {
        multitenancy: { enabled: true },
        auth: {
          type: 'dummy',
        },
      },
    };

    // Render the component
    const { container, queryByText } = render(<AccountNavButton {...props} />);

    // Find the popover component by data-test-subj
    const popoverElement = container.querySelector('[data-test-subj="account-popover"]');

    // Assert that the popover component is rendered
    expect(popoverElement).toBeDefined();

    // Now, simulate a click event on the button to open the popover
    fireEvent.click(popoverElement!);

    const tenantName = queryByText('example_tenant');
    expect(tenantName).not.toBeNull();

    const tenantSwitch = queryByText('Switch tenants');
    expect(tenantSwitch).not.toBeNull();
  });

  test('Does not render "switch-tenants" and "tenant-name" when multi-tenancy is disabled', () => {
    const props = {
      coreStart: {},
      isInternalUser: true,
      username: 'example_user',
      tenant: 'example_tenant',
      config: {
        multitenancy: { enabled: false },
        auth: {
          type: 'dummy',
        },
      },
    };

    // Render the component
    const { container, queryByText } = render(<AccountNavButton {...props} />);

    // Find the popover component by data-test-subj
    const popoverElement = container.querySelector('[data-test-subj="account-popover"]');

    // Assert that the popover component is rendered
    expect(popoverElement).toBeDefined();

    // Now, simulate a click event on the button to open the popover
    fireEvent.click(popoverElement!);

    const tenantName = queryByText('example_tenant');
    expect(tenantName).toBeNull();

    const tenantSwitch = queryByText('Switch tenants');
    expect(tenantSwitch).toBeNull();
  });
});

describe('Reload window after tenant switch', () => {
  const originalLocation = window.location;
  const mockSetWindowHref = jest.fn();
  let pathname: string = '';
  beforeAll(() => {
    pathname = '/app/myapp';
    Object.defineProperty(window, 'location', {
      value: {
        get pathname() {
          return pathname;
        },
        get href() {
          return '/app/dashboards?security_tenant=admin_tenant';
        },
        set href(value: string) {
          mockSetWindowHref(value);
        },
      },
    });
  });

  afterAll(() => {
    window.location = originalLocation;
  });

  it('should remove the tenant query parameter before reloading', () => {
    pathname = '/app/pathname-only';
    reloadAfterTenantSwitch();
    expect(mockSetWindowHref).toHaveBeenCalledWith(pathname);
  });
});

describe('Clear lastUrls after tenant switch', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('should clear out keys with a lastUrl prefix', () => {
    window.sessionStorage.setItem('lastUrl:dashboard', '/dashboard1');
    window.sessionStorage.setItem('lastUrl:otherApp', '/otherApp');
    window.sessionStorage.setItem('somethingElse:here', '/random');
    const mockRemoveItem = jest.spyOn(Object.getPrototypeOf(window.sessionStorage), 'removeItem');
    reloadAfterTenantSwitch();
    expect(mockRemoveItem).toHaveBeenCalledWith('lastUrl:dashboard');
    expect(mockRemoveItem).toHaveBeenCalledWith('lastUrl:otherApp');
    expect(mockRemoveItem).toHaveBeenCalledTimes(2);
  });

  it('should not clear out keys without a lastUrl prefix', () => {
    window.sessionStorage.setItem('somethingElse:here', '/random');
    const mockRemoveItem = jest.spyOn(Object.getPrototypeOf(window.sessionStorage), 'removeItem');

    reloadAfterTenantSwitch();
    expect(mockRemoveItem).toHaveBeenCalledTimes(0);
  });
});
