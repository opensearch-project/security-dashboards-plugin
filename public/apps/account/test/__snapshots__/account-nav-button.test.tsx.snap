// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account navigation button renders 1`] = `
<EuiHeaderSectionItemButton
  id="user-icon-btn"
>
  <EuiPopover
    anchorPosition="downCenter"
    button={
      <EuiAvatar
        name="user1"
      />
    }
    closePopover={[Function]}
    data-test-subj="account-popover"
    display="inlineBlock"
    hasArrow={true}
    id="actionsMenu"
    isOpen={false}
    onClick={[Function]}
    ownFocus={true}
    panelPaddingSize="s"
  >
    <EuiContextMenuPanel
      hasFocus={true}
      items={Array []}
    >
      <div
        style={
          Object {
            "maxWidth": "256px",
          }
        }
      >
        <EuiFlexGroup
          gutterSize="s"
        >
          <EuiFlexItem
            grow={null}
          >
            <EuiAvatar
              name="user1"
            />
          </EuiFlexItem>
          <EuiFlexItem>
            <EuiListGroup
              gutterSize="none"
            >
              <EuiListGroupItem
                key="username"
                label={
                  <EuiText
                    size="s"
                  >
                    <h5>
                      user1
                    </h5>
                  </EuiText>
                }
                wrapText={true}
              />
            </EuiListGroup>
            <EuiListGroupItem
              color="subdued"
              key="tenant"
              label={
                <EuiText
                  id="tenantName"
                  size="xs"
                >
                  tenant1
                </EuiText>
              }
            />
          </EuiFlexItem>
        </EuiFlexGroup>
        <EuiHorizontalRule
          margin="xs"
        />
        <EuiButtonEmpty
          data-test-subj="view-roles-and-identities"
          onClick={[Function]}
          size="xs"
        >
          View roles and identities
        </EuiButtonEmpty>
        <EuiHorizontalRule
          margin="xs"
        />
        <EuiButtonEmpty
          data-test-subj="switch-tenants"
          onClick={[Function]}
          size="xs"
        >
          Switch tenants
        </EuiButtonEmpty>
        <EuiHorizontalRule
          margin="xs"
        />
        <EuiButtonEmpty
          data-test-subj="reset-password"
          onClick={[Function]}
          size="xs"
        >
          Reset password
        </EuiButtonEmpty>
        <LogoutButton
          authType="dummy"
          divider={
            <EuiHorizontalRule
              margin="xs"
            />
          }
          http={1}
        />
      </div>
    </EuiContextMenuPanel>
  </EuiPopover>
</EuiHeaderSectionItemButton>
`;
